### PR TITLE
Raster singleband legend precision

### PR DESCRIPTION
**Pull request by elpaso**
Followup #39058 and #39071 I don't really know if this qualifies as a bugfix or a feature.

immagine

CC: @gioman @PedroVenancio

### CODE REVIEW

File: src/gui/qgsguiutils.cpp [hunk 1]
```diff
@@ -275,6 +275,48 @@ namespace QgsGuiUtils
...
+      {
+        precision = 9;
+        break;
+      }
+      case Qgis::DataType::Float64:
+      {
+        precision = 17;
+        break;
+      }
+      default:
```
Comment (nyalldawson) [inline src/gui/qgsguiutils.cpp:312]: Can you drop the default clause and list all enum values explicitly here? Then we will get a build warning if new types are added and the developer knows they have to update this method accordingly

**Comment by nyalldawson** [comment]
Regarding the UI: could we avoid the checkbox and instead set the clear value for the spin to use a "default" string?

I'm happy for this to be classed as a bug fix.

**Comment by elpaso** [comment]
> Regarding the UI: could we avoid the checkbox and instead set the clear value for the spin to use a "default" string?

@nyalldawson  Actually, I removed the checkbox completely and set the max precision of the spinbox to the max significant digits. It looks simpler and more user friendly this way.

**Comment by nyalldawson** [comment]
@elpaso got a screenshot of that?

**Comment by elpaso** [comment]
> @elpaso got a screenshot of that?

Image updated

File: src/gui/raster/qgscolorrampshaderwidget.cpp [hunk 2]
```diff
@@ -118,6 +120,18 @@ void QgsColorRampShaderWidget::setRasterDataProvider( QgsRasterDataProvider *dp
...
+  if ( mRasterDataProvider )
+  {
+    const int maxDigits { QgsGuiUtils::significantDigits( mRasterDataProvider->dataType( mBand ) ) };
+    // Get current value
+    const int currentPrecision { mLabelPrecisionSpinBox->value() };
+    mLabelPrecisionSpinBox->setMaximum( maxDigits );
+    if ( currentPrecision > maxDigits )
+    {
+      mLabelPrecisionSpinBox->setValue( maxDigits );
+    }
```
Comment (nyalldawson) [inline src/gui/raster/qgscolorrampshaderwidget.cpp:133]: Is this needed? Won't setting the maximum for the spin box already handle this situation?

Comment (elpaso) [reply src/gui/raster/qgscolorrampshaderwidget.cpp:133]: I'm not sure: does it? I'll check. There is also the question if valueChanged is triggered (it should).

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**